utils/visualizer.py

```diff
@@ -1,62 +1,635 @@
+        (file: varname +"_scatter_condm_err.png")
+    create_scatter_plot_variable(self, varname, true_values, predicted_values, iepoch=None, save_plot=True)
+        Creates scatter plots for true_values and predicted values of variable varname at iepoch.
+        (file: varname + ".png")
+    create_error_histogram_plot_nodes(self, varname, true_values, predicted_values, iepoch=None, save_plot=True)
+        Creates error histogram plot for the true and predicted values of variable varname at iepoch.[node level]
+        (file: varname+"_error_hist1d_"+ str(iepoch).zfill(4)+ ".png" or varname+"_error_hist1d.png" )
+    create_scatter_plot_nodes_vec(self, varname, true_values, predicted_values, iepoch=None, save_plot=True):
+        Creates scatter plots for true and predicted values of vector variable varname.
+        (file: varname+"_"+ str(iepoch).zfill(4) + ".png" or varname+".png" )
+
+    #create plots for all heads
+    plot_history(self,trainlib,vallib,testlib,tasklib,tasklib_vali,tasklib_test,tasklib_nodes,tasklib_vali_nodes,tasklib_test_nodes,task_weights,task_names,):
+        Save history of losses to file and plot.
+    create_scatter_plots(self, true_values, predicted_values, output_names=None, iepoch=None)
+        Creates scatter plots for all head predictions. One plot for each head
+    create_plot_global(self, true_values, predicted_values, output_names=None)
+        Creates global analysis for all head predictons, e.g., scatter/condmean/error pdf plot. One plot for each head
     """
 
-    def __init__(self, model_with_config_name: str):
+    def __init__(
+        self, model_with_config_name: str, node_feature: [], num_heads=1, head_dims=[1]
+    ):
         self.true_values = []
         self.predicted_values = []
         self.model_with_config_name = model_with_config_name
+        self.node_feature = node_feature
+        self.num_nodes = len(node_feature[0])
+        self.num_heads = num_heads
+        self.head_dims = head_dims
+
+    def __hist2d_contour(self, data1, data2):
+        hist2d_pasr, xedge_pasr, yedge_pasr = np.histogram2d(
+            np.hstack(data1), np.hstack(data2), bins=50
+        )
+        xcen_pasr = 0.5 * (xedge_pasr[0:-1] + xedge_pasr[1:])
+        ycen_pasr = 0.5 * (yedge_pasr[0:-1] + yedge_pasr[1:])
+        BCTY_pasr, BCTX_pasr = np.meshgrid(ycen_pasr, xcen_pasr)
+        hist2d_pasr = hist2d_pasr / np.amax(hist2d_pasr)
+        return BCTX_pasr, BCTY_pasr, hist2d_pasr
+
+    def __err_condmean(self, data1, data2, weight=1.0):
+        errabs = np.abs(np.hstack(data1) - np.hstack(data2)) * weight
+        hist2d_pasr, xedge_pasr, yedge_pasr = np.histogram2d(
+            np.hstack(data1), errabs, bins=50
+        )
+        xcen_pasr = 0.5 * (xedge_pasr[0:-1] + xedge_pasr[1:])
+        ycen_pasr = 0.5 * (yedge_pasr[0:-1] + yedge_pasr[1:])
+        hist2d_pasr = hist2d_pasr / np.amax(hist2d_pasr)
+        mean1d_cond = np.dot(hist2d_pasr, ycen_pasr) / np.sum(hist2d_pasr, axis=1)
```

Comment:
`RuntimeWarning: invalid value encountered in true_divide`

Reply:
It is caused by divide by zero I think. Added a small number in e312a340. It should be gone now.